### PR TITLE
Multiple Fixes to dApp reward retrieval

### LINKED ISSUE #1683: First reward retrieval should be possible from within the voter-dapp

### LINKED ISSUE #1695: Voter dApp should not allow voters who voted incorrectly to claim rewards
`batchRetrieveRewards` is callable by anyone to retrieve rewards for anyone. The dApp enables users to call this on their own address, but should block them from doing so if they do not have any outstanding rewards.

### PR DESCRIPTION
**Pull request by mrice32**
1. Fixes #1695.
2. Fixes #1683.

I've tested this locally in ganache, but likely haven't covered *all* edge cases. Please suggest test cases that should be run if you can think of any interesting ones.

Note: this does not handle prevent users from erroneously attempting to retrieve rewards that are expired. However, the default reward retrieval lookback was updated to not show expired rewards by default. This should be fixed in a separate PR.

### CODE REVIEW

File: voter-dapp/src/RetrieveRewards.js [hunk 1]
```diff
@@ -154,13 +146,49 @@ function RetrieveRewards({ votingAccount }) {
...
+    // Pulls down the price for every resolved request. If the request wasn't resolved, the resolved price isn't
+    // queried and the field is left undefined.
+    let done = true;
+    for (let i = 0; i < priceRequests.length; i++) {
+      const priceRequest = priceRequests[i];
+      const status = statuses[i];
+      priceRequest.lastRound = status.lastVotingRound;
+      if (status.status === PriceRequestStatusEnum.RESOLVED) {
+        priceRequest.resolvedPrice = call("Voting", "getPrice", priceRequest.identifier, priceRequest.time, {
+          from: governorAddress
```
Comment (chrismaree) [inline voter-dapp/src/RetrieveRewards.js:179]: question: The `getPrice` function is a public view function. Why do you need this `from: governorAddress` here?

Comment (mrice32) [reply voter-dapp/src/RetrieveRewards.js:179]: It can only be called by approved financial contracts. It will revert if called by the user's EOA. But for `.calls`, you can just tell the node to simulate the call as if it were called from a different address.

I'll add a comment to explain.

File: voter-dapp/src/RetrieveRewards.js [hunk 2]
```diff
@@ -14,13 +15,13 @@ function getOrCreateObj(containingObj, field) {
...
 
-function useRetrieveRewardsTxn(retrievedRewardsEvents, revealedVoteEvents, priceResolvedEvents, votingAccount) {
+function useRetrieveRewardsTxn(retrievedRewardsEvents, revealedVoteEvents, priceRequests, votingAccount) {
   const { drizzle, useCacheSend } = drizzleReactHooks.useDrizzle();
   const { web3 } = drizzle;
 
   const { send, status } = useCacheSend("Voting", "retrieveRewards");
 
-  if (retrievedRewardsEvents === undefined || revealedVoteEvents === undefined || priceResolvedEvents === undefined) {
+  if (retrievedRewardsEvents === undefined || revealedVoteEvents === undefined || priceRequests == null) {
```
Comment (nicholaspai) [inline voter-dapp/src/RetrieveRewards.js:24]: missing third "="

Comment (mrice32) [reply voter-dapp/src/RetrieveRewards.js:24]: This was intentional so it would compare as equal to `null` or `undefined`. Is there a better way to signal that? Would `!priceRequests` be preferable?

Comment (mrice32) [reply voter-dapp/src/RetrieveRewards.js]: Changed to `!priceRequests`. PTAL.

File: voter-dapp/src/RetrieveRewards.js (hunk 1 above)
Comment (mrice32) [reply voter-dapp/src/RetrieveRewards.js:181]: Added a comment. PTAL.

**Comment by mrice32** [comment]
@nicholaspai @chrismaree responded to your comments. PTAL.

File: voter-dapp/src/RetrieveRewards.js [hunk 3]
```diff
@@ -43,24 +44,26 @@ function useRetrieveRewardsTxn(retrievedRewardsEvents, revealedVoteEvents, price
...
+      const event = revealedVoteEvents[i];
+      const priceRequest = priceRequests[i];
       const voteState = getVoteState(event.returnValues.identifier, event.returnValues.time);
       const revealRound = event.returnValues.roundId.toString();
+      const votedPrice = event.returnValues.price.toString();
 
-      if (!voteState.retrievedRewards && voteState.priceResolutionRound === revealRound) {
+      if (
+        !voteState.retrievedRewards &&
+        priceRequest.lastRound.toString() === revealRound &&
```
Comment (nicholaspai) [inline voter-dapp/src/RetrieveRewards.js:60]: Why do you have to make sure the last round was the Reveal round? Aren't rewards able to be retrieved once the `votedPrice === priceRequest.resolvedPrice.toString()` (i.e. once there exists a resolved price) ?

File: voter-dapp/src/RetrieveRewards.js [hunk 4]
```diff
@@ -14,13 +15,13 @@ function getOrCreateObj(containingObj, field) {
...
 
-function useRetrieveRewardsTxn(retrievedRewardsEvents, revealedVoteEvents, priceResolvedEvents, votingAccount) {
+function useRetrieveRewardsTxn(retrievedRewardsEvents, revealedVoteEvents, priceRequests, votingAccount) {
   const { drizzle, useCacheSend } = drizzleReactHooks.useDrizzle();
   const { web3 } = drizzle;
 
   const { send, status } = useCacheSend("Voting", "retrieveRewards");
 
-  if (retrievedRewardsEvents === undefined || revealedVoteEvents === undefined || priceResolvedEvents === undefined) {
+  if (retrievedRewardsEvents === undefined || revealedVoteEvents === undefined || !priceRequests) {
```
Comment (nicholaspai) [inline voter-dapp/src/RetrieveRewards.js:24]: nit: it seems to me all of these checks are doing the same thing (whether some variable is `undefined`) but why do you use `!priceRequests` versus `revealedVoteEvents === undefined`. I think it'd be a bit cleaner to use all "!"'s

File: voter-dapp/src/RetrieveRewards.js [hunk 5]
```diff
@@ -14,13 +15,13 @@ function getOrCreateObj(containingObj, field) {
   return containingObj[field];
 }
 
-function useRetrieveRewardsTxn(retrievedRewardsEvents, revealedVoteEvents, priceResolvedEvents, votingAccount) {
+function useRetrieveRewardsTxn(retrievedRewardsEvents, revealedVoteEvents, priceRequests, votingAccount) {
```
Comment (nicholaspai) [inline voter-dapp/src/RetrieveRewards.js:18]: Maybe this is just me, but I was a bit confused by the name `priceRequests`. I didn't realize that `priceRequests` and `revealedVoteEvents` were 1-to-1 mapped. I think it'd be a bit cleaner if you collapsed all of the data into one object OR you indicated that `priceRequests` was some sort of `revealedVote`-mapping-to-`priceRequests`. 

Additionally, is `retrievedRewardsEvents` used anywhere in this method?

Comment (nicholaspai) [reply voter-dapp/src/RetrieveRewards.js:18]: >Additionally, is retrievedRewardsEvents used anywhere in this method?

Ah I see it is used. nevermind

File: voter-dapp/src/RetrieveRewards.js [hunk 6]
```diff
@@ -117,7 +117,7 @@ function RetrieveRewards({ votingAccount }) {
       return MAX_UINT_VAL;
     } else {
       // This section will produce an array of roundIds that should be queried for unclaimed rewards.
-      const defaultLookback = 10; // Default lookback is 10 rounds.
+      const defaultLookback = 7; // Default lookback is 7 rounds (2 weeks).
```
Comment (nicholaspai) [inline voter-dapp/src/RetrieveRewards.js:120]: OOC what was your thinking behind this hard coded value?

File: voter-dapp/src/RetrieveRewards.js (hunk 4 above)
Comment (mrice32) [reply voter-dapp/src/RetrieveRewards.js:24]: Yeah, this is a drizzle thing.

The first two come from naked drizzle calls. Naked drizzle calls return `undefined` to mean "result hasn't come back yet" and `null` to mean "error or empty result". The first two should generally never have `null`, so I check for undefined specifically, since I only want to catch "hasn't come back yet". I don't want this to silently ignore real errors or empty values coming back from these calls.

The last one should really only be `null`, since we dictate the return values for that one (not a naked drizzle call). I should probably explicitly check for null, tbh. I'll change that.

File: voter-dapp/src/RetrieveRewards.js (hunk 5 above)
Comment (mrice32) [reply voter-dapp/src/RetrieveRewards.js:18]: Yeah, a lot of this is just because I'm trying to retrofit this method to a new way of computing these values.

Collapsing retrievedRewardsEvents and revealedVoteEvents is what a lot of this method is doing. There are a lot of subtleties in doing that -- one being that revealedVoteEvents can have multiple votes on the same price request in the case that the earlier votes don't hit the GAT. Some revealed votes don't exist in retrieved rewards (for obvious reasons). Merging the two requires you to create a completely distinct data structure that can handle all the subtle partial overlaps. That's a lot of what this method is trying to handle (there used to be a third set of overlapping events, too, which made it more necessary.

I agree that having priceRequests be 1:1 mapped with revealed votes is a little confusing. I could technically make them part of the same array pretty easily. Wasn't sure if it would be more confusing to dump price request data into the revealedVoteEvent array. Do you think mutating those event structs would make things simpler? I'm happy to do that -- it wouldn't change the structure too much.

Thoughts?

File: voter-dapp/src/RetrieveRewards.js (hunk 6 above)
Comment (mrice32) [reply voter-dapp/src/RetrieveRewards.js:120]: Rewards expire after two weeks (7 voting periods). I can add a comment.

File: voter-dapp/src/RetrieveRewards.js (hunk 3 above)
Comment (mrice32) [reply voter-dapp/src/RetrieveRewards.js:60]: This is to handle this case:
- You reveal the correct vote in round 5.
- The price request doesn't pass the GAT in round 5.
- You reveal the incorrect vote or don't reveal in round 6.
- The price request passes the GAT in round 6 resolving to your vote in round 5.

Put simply, the right vote at the wrong time means you don't get rewards :).

File: voter-dapp/src/RetrieveRewards.js (hunk 4 above)
Comment (mrice32) [reply voter-dapp/src/RetrieveRewards.js]: Note: I changed the last one to null. Do you think that makes sense?

Comment (mrice32) [reply voter-dapp/src/RetrieveRewards.js]: @nicholaspai I updated the third object that we control to be undefined if the call hasn't finished resolving yet to make it work like the regular drizzle calls. This allowed me to make them all the same `=== undefined`. PTAL.

File: voter-dapp/src/RetrieveRewards.js (hunk 3 above)
Comment (mrice32) [reply voter-dapp/src/RetrieveRewards.js:62]: Added a comment to this effect.

File: voter-dapp/src/RetrieveRewards.js (hunk 2 above)
Comment (nicholaspai) [reply voter-dapp/src/RetrieveRewards.js]: cool makes sense

File: voter-dapp/src/RetrieveRewards.js (hunk 5 above)
Comment (nicholaspai) [reply voter-dapp/src/RetrieveRewards.js:18]: I do think putting them into the same structure is more sustainable. This way no one can unintentionally cause the two arrays to be misaligned

File: voter-dapp/src/RetrieveRewards.js (hunk 4 above)
Comment (nicholaspai) [reply voter-dapp/src/RetrieveRewards.js]: >The first two come from naked drizzle calls. Naked drizzle calls return undefined to mean "result hasn't come back yet" and null to mean "error or empty result". 

This is horrible lol. I understand though

File: voter-dapp/src/RetrieveRewards.js (hunk 3 above)
Comment (nicholaspai) [reply voter-dapp/src/RetrieveRewards.js:62]: So `lastRound` is when the price request was resolved, and `revealRound` is when the user revealed?

I think I got confused here because I though `revealRound` was opposite of `commitRound`, but I think looking at it now that `revealRound` is actually the round that a price was revealed.

File: voter-dapp/src/RetrieveRewards.js (hunk 6 above)
Comment (nicholaspai) [reply voter-dapp/src/RetrieveRewards.js]: Makes sense

File: voter-dapp/src/RetrieveRewards.js (hunk 3 above)
Comment (mrice32) [reply voter-dapp/src/RetrieveRewards.js:62]: Yep, that's correct. Let me know if there's a comment I can add that would help make this clearer.

File: voter-dapp/src/RetrieveRewards.js (hunk 4 above)
Comment (mrice32) [reply voter-dapp/src/RetrieveRewards.js]: > This is horrible lol. I understand though
lol yup

File: voter-dapp/src/RetrieveRewards.js (hunk 5 above)
Comment (mrice32) [reply voter-dapp/src/RetrieveRewards.js]: @nicholaspai done. PTAL.

**Comment by mrice32** [comment]
> I responded to your comments, looks good. Just want to make sure you tested the various flows:
> 
> * User uses 2key contract to vote and retrieve rewards
> * User uses normal hot wallet to vote and retrieve rewards
> 
> Optionally:
> 
> * User votes with CLI and retrieves rewards with dApp
> * User votes with dApp and retrieves rewards with dApp
> * User votes with CLI using 2Key and retrieves rewards with dApp
> * User votes with dApp using 2Key and retrieves rewards with dApp
> 
> These different branches are why supporting both the CLI and the dApp becomes unsustainable eventually...

I tested the first two as well as a lot of weird rewards edge cases (like votes not hitting GAT). The latter ones, I think I'll leave for our pre-release QA testing. But these should definitely all be on the pre-release list.
